README.md
# unilist

[![Build Status](https://semaphoreci.com/api/v1/sourcefoundry/unilist/branches/master/badge.svg)](https://semaphoreci.com/sourcefoundry/unilist)
[![Build status](https://ci.appveyor.com/api/projects/status/c79jc4vl9jbiki0g/branch/master?svg=true)](https://ci.appveyor.com/project/chrissimpkins/unilist/branch/master)
[![Go Report Card](https://goreportcard.com/badge/github.com/source-foundry/unilist)](https://goreportcard.com/report/github.com/source-foundry/unilist)

unilist is a command line executable that writes lists of Unicode code points out to the standard output stream.  It supports space-, comma-, and newline-delimited lists of Unicode code points that are defined with inclusion and exclusion arguments.  The application was developed to support the creation of Unicode code point lists for use with font subsetting tools.

```
# Example of Basic Latin - Latin Extended A sets, excluding U+0080-009F, comma-delimited

$ unilist -c 0020-017F x0080-009F
```

**Output**:
U+0020,U+0021,U+0022,U+0023,U+0024,U+0025,U+0026,U+0027,U+0028,U+0029,U+002A,U+002B,U+002C,U+002D,U+002E,U+002F,U+0030,U+0031,U+0032,U+0033,U+0034,U+0035,U+0036,U+0037,U+0038,U+0039,U+003A,U+003B,U+003C,U+003D,U+003E,U+003F,U+0040,U+0041,U+0042,U+0043,U+0044,U+0045,U+0046,U+0047,U+0048,U+0049,U+004A,U+004B,U+004C,U+004D,U+004E,U+004F,U+0050,U+0051,U+0052,U+0053,U+0054,U+0055,U+0056,U+0057,U+0058,U+0059,U+005A,U+005B,U+005C,U+005D,U+005E,U+005F,U+0060,U+0061,U+0062,U+0063,U+0064,U+0065,U+0066,U+0067,U+0068,U+0069,U+006A,U+006B,U+006C,U+006D,U+006E,U+006F,U+0070,U+0071,U+0072,U+0073,U+0074,U+0075,U+0076,U+0077,U+0078,U+0079,U+007A,U+007B,U+007C,U+007D,U+007E,U+007F,U+00A0,U+00A1,U+00A2,U+00A3,U+00A4,U+00A5,U+00A6,U+00A7,U+00A8,U+00A9,U+00AA,U+00AB,U+00AC,U+00AD,U+00AE,U+00AF,U+00B0,U+00B1,U+00B2,U+00B3,U+00B4,U+00B5,U+00B6,U+00B7,U+00B8,U+00B9,U+00BA,U+00BB,U+00BC,U+00BD,U+00BE,U+00BF,U+00C0,U+00C1,U+00C2,U+00C3,U+00C4,U+00C5,U+00C6,U+00C7,U+00C8,U+00C9,U+00CA,U+00CB,U+00CC,U+00CD,U+00CE,U+00CF,U+00D0,U+00D1,U+00D2,U+00D3,U+00D4,U+00D5,U+00D6,U+00D7,U+00D8,U+00D9,U+00DA,U+00DB,U+00DC,U+00DD,U+00DE,U+00DF,U+00E0,U+00E1,U+00E2,U+00E3,U+00E4,U+00E5,U+00E6,U+00E7,U+00E8,U+00E9,U+00EA,U+00EB,U+00EC,U+00ED,U+00EE,U+00EF,U+00F0,U+00F1,U+00F2,U+00F3,U+00F4,U+00F5,U+00F6,U+00F7,U+00F8,U+00F9,U+00FA,U+00FB,U+00FC,U+00FD,U+00FE,U+00FF,U+0100,U+0101,U+0102,U+0103,U+0104,U+0105,U+0106,U+0107,U+0108,U+0109,U+010A,U+010B,U+010C,U+010D,U+010E,U+010F,U+0110,U+0111,U+0112,U+0113,U+0114,U+0115,U+0116,U+0117,U+0118,U+0119,U+011A,U+011B,U+011C,U+011D,U+011E,U+011F,U+0120,U+0121,U+0122,U+0123,U+0124,U+0125,U+0126,U+0127,U+0128,U+0129,U+012A,U+012B,U+012C,U+012D,U+012E,U+012F,U+0130,U+0131,U+0132,U+0133,U+0134,U+0135,U+0136,U+0137,U+0138,U+0139,U+013A,U+013B,U+013C,U+013D,U+013E,U+013F,U+0140,U+0141,U+0142,U+0143,U+0144,U+0145,U+0146,U+0147,U+0148,U+0149,U+014A,U+014B,U+014C,U+014D,U+014E,U+014F,U+0150,U+0151,U+0152,U+0153,U+0154,U+0155,U+0156,U+0157,U+0158,U+0159,U+015A,U+015B,U+015C,U+015D,U+015E,U+015F,U+0160,U+0161,U+0162,U+0163,U+0164,U+0165,U+0166,U+0167,U+0168,U+0169,U+016A,U+016B,U+016C,U+016D,U+016E,U+016F,U+0170,U+0171,U+0172,U+0173,U+0174,U+0175,U+0176,U+0177,U+0178,U+0179,U+017A,U+017B,U+017C,U+017D,U+017E,U+017F

## Contents

- [Installation](#installation)
- [Usage](#usage)
- [Issues](#issues)
- [Contributing](#contributing)
- [License](#license)

## Installation

unilist is developed in Go and compiled to the command line executable `unilist` (`unilist.exe` on Windows). A variety of cross-compiled binaries are available for use on Linux, macOS, and Windows systems, or you can download the source and compile the application yourself. Instructions for both approaches follow.

## Installation Approaches

### Approach 1: Install the pre-compiled binary executable file

Download the latest compiled release file for your operating system and architecture from [the Releases page](https://github.com/source-foundry/unilist/releases/latest).

#### Linux / macOS

Unpack the tar.gz archive and move the `unilist` executable file to a directory on your system PATH (e.g. `/usr/local/bin`).  Execute this move with the following command in the root of the unpacked archive:

```
$ mv unilist /usr/local/bin/unilist
```

There are no dependencies contained in the archive.  You can delete all downloaded archive files after the above step.

#### Windows

Unpack the zip archive and move the `unilist.exe` executable file to a directory on your system PATH. See [details here](https://stackoverflow.com/questions/4822400/register-an-exe-so-you-can-run-it-from-any-command-line-in-windows) for more information about how to do this.

There are no dependencies contained in the archive.  You can delete all downloaded archive files after the above step.

### Approach 2: Compile from the source code and install

You must install the Go programming language (which includes the `go` tool) to compile the project from source.  Follow the [instructions on the Go download page](https://golang.org/dl/) for your platform. 

Once you have installed Go and configured your settings so that Go executables are installed on your system PATH, use the following command to (1) pull the master branch of the repository; (2) compile the `unilist` executable from source for your platform/architecture configuration; (3) install the executable on your system:

```
$ go get -u github.com/source-foundry/unilist
```

## Uninstall

The installation includes a single executable binary file.  If you installed with `go get` or added one of the pre-compiled binaries on your system `$PATH` on *.nix systems, you can uninstall with:

```
$ rm $(which unilist)
```

## Usage

```
$ unilist [option] [arg]
```

where `arg` is a single Unicode code point hexadecimal value or a range of hexadecimal values as defined below.

#### How to define included Unicode code points

##### Individual Unicode code points

```
$ unilist 0020
```

The syntax for single Unicode code point values is the hexadecimal value without prefix characters (i.e., do not use `U+`, `u+` or any other commonly used prefix characters to indicate that these are Unicode code point values) and must be at least 4 digits in length.  An example of a valid definition of the space glyph is `0020`.

##### A range of Unicode code points

```
$ unilist 0020-0025
```

The syntax for Unicode code point ranges is the start hexadecimal value, a dash `-`, the end hexadecimal value.  For example, a request to include Unicode code points with values `0020`, `0021`, `0022`, `0023`, `0024`, and `0025` should appear like this on the command line: `0020-0025`.

#### How to define excluded Unicode code points

unilist supports individual code point and code point range exclusions from the list of values with a case-insensitive `x` glyph at the first position of the command line argument.  The hexadecimal value in the argument must use the format described in the section above.

##### Individual Unicode code points

```
$ unilist 0020-0025 x0021
```

Add the `x` at the first character position of your Unicode code point argument to define it for exclusion from the list.  

For example, use the argument `x0021` to exclude the Unicode code point `0021` from the list.

##### A range of Unicode code points

```
$ unilist 0020-0025 x0021-0022
```

Add the `x` at the first character position of your Unicode code point range argument to define the range of Unicode code point values for exclusion from the list.  

For example, use the argument `x0021-0022` to exclude the Unicode code points `0021` and `0022` from the list.  Note that this is a shortcut for behavior that can be specified with multiple individual code point definitions like this: `x0021 x0022`.

### Default Behavior

The unilist default is a space-delimited list of Unicode code point values with the format `U+[value]`.  

Modify the delimiter between Unicode code point values with the use of an option in your command as described below.

### Options

```
  -c, --comma          Use comma delimiter
  -h, --help           Application help
  -n, --newline        Use newline delimiter
      --usage          Application usage
  -v, --version        Application version
```

## Issues

Please [file an issue report](https://github.com/source-foundry/unilist/issues/new) on the repository for any problems that arise with use.

## Contributing

Contributions to this project are welcomed. Please submit your changes as a pull request on this repository.

## License

Licensed under the [MIT License](LICENSE)

CHANGELOG.md
## Changelog

### v0.3.0

- added Unicode code point argument length checks for range arguments
- improved in-application help text
- refactored error names in parser.go main function
- go fmt source formatting modifications
- added Windows CI testing on Appveyor CI
- added Linux CI testing on Semaphore CI

### v0.2.0

- refactored command line Unicode hexadecimal code argument parser to new source file `parser.go`
- refactored string utility functions to new source file `stringutils.go`

### v0.1.0

- initial release

parser_test.go
package main

import (
	"testing"
)

//
//  INCLUDE VALUE TESTS
//

func TestGetUnicodeValueSingleValidIncludeValue(t *testing.T) {
	testArg := "0020"
	resSlice, err := getUnicodeValueSlice(testArg)
	if err != nil {
		t.Errorf("[FAIL] Expected no error and received error %v", err)
	}
	if len(resSlice) != 1 {
		t.Errorf("[FAIL] Expected one return value in the slice, received %d", len(resSlice))
	}
	if resSlice[0] != "U+0020" {
		t.Errorf("[FAIL] Did not expect returned value %s", resSlice[0])
	}
}

func TestGetUnicodeValueRangeValidIncludeValues(t *testing.T) {
	testArg := "0020-0022"
	resSlice, err := getUnicodeValueSlice(testArg)
	if err != nil {
		t.Errorf("[FAIL] Expected no error and received error %v", err)
	}
	if len(resSlice) != 3 {
		t.Errorf("[FAIL] Expected one return value in the slice, received %d", len(resSlice))
	}
	if resSlice[0] != "U+0020" {
		t.Errorf("[FAIL] Did not expect returned value %s", resSlice[0])
	}
	if resSlice[1] != "U+0021" {
		t.Errorf("[FAIL] Did not expect returned value %s", resSlice[0])
	}
	if resSlice[2] != "U+0022" {
		t.Errorf("[FAIL] Did not expect returned value %s", resSlice[0])
	}
}

func TestGetUnicodeValueSingleInvalidIncludeValueNotHexadecimal(t *testing.T) {
	testArg := "ZZZZ"
	_, err := getUnicodeValueSlice(testArg)
	if err == nil {
		t.Errorf("[FAIL] Expected error for invalid hexadecimal string and received no error")
	}
}

func TestGetUnicodeValueSingleInvalidIncludeIncorrectLengthHexadecimal(t *testing.T) {
	testArg := "20"
	_, err := getUnicodeValueSlice(testArg)
	if err == nil {
		t.Errorf("[FAIL] Expected error for invalid length hexadecimal string and received no error")
	}
}

func TestGetUnicodeValueRangeInvalidIncludeFirstValueNotHexadecimal(t *testing.T) {
	testArg := "ZZZZ-0020"
	_, err := getUnicodeValueSlice(testArg)
	if err == nil {
		t.Errorf("[FAIL] Expected error for invalid hexadecimal string and received no error")
	}
}

func TestGetUnicodeValueRangeInvalidIncludeLastValueNotHexadecimal(t *testing.T) {
	testArg := "0020-ZZZZ"
	_, err := getUnicodeValueSlice(testArg)
	if err == nil {
		t.Errorf("[FAIL] Expected error for invalid hexadecimal string and received no error")
	}
}

func TestGetUnicodeValueRangeInvalidIncludeTooShortStart(t *testing.T) {
	testArg := "20-0030"
	_, err := getUnicodeValueSlice(testArg)
	if err == nil {
		t.Errorf("[FAIL] Expected error for invalid hexadecimal string and received no error")
	}
}

func TestGetUnicodeValueRangeInvalidIncludeTooShortEnd(t *testing.T) {
	testArg := "0020-30"
	_, err := getUnicodeValueSlice(testArg)
	if err == nil {
		t.Errorf("[FAIL] Expected error for invalid hexadecimal string and received no error")
	}
}

//
// EXCLUDE VALUE TESTS
//

func TestGetUnicodeValueSingleValidExcludeValue(t *testing.T) {
	testArg := "x0020"
	resSlice, err := getUnicodeValueSlice(testArg)
	if err != nil {
		t.Errorf("[FAIL] Expected no error and received error %v", err)
	}
	if len(resSlice) != 1 {
		t.Errorf("[FAIL] Expected one return value in the slice, received %d", len(resSlice))
	}
	if resSlice[0] != "U+0020" {
		t.Errorf("[FAIL] Did not expect returned value %s", resSlice[0])
	}
}

func TestGetUnicodeValueRangeValidExcludeValues(t *testing.T) {
	testArg := "x0020-0022"
	resSlice, err := getUnicodeValueSlice(testArg)
	if err != nil {
		t.Errorf("[FAIL] Expected no error and received error %v", err)
	}
	if len(resSlice) != 3 {
		t.Errorf("[FAIL] Expected one return value in the slice, received %d", len(resSlice))
	}
	if resSlice[0] != "U+0020" {
		t.Errorf("[FAIL] Did not expect returned value %s", resSlice[0])
	}
	if resSlice[1] != "U+0021" {
		t.Errorf("[FAIL] Did not expect returned value %s", resSlice[0])
	}
	if resSlice[2] != "U+0022" {
		t.Errorf("[FAIL] Did not expect returned value %s", resSlice[0])
	}
}

func TestGetUnicodeValueSingleInvalidExcludeValueNotHexadecimal(t *testing.T) {
	testArg := "xZZZZ"
	_, err := getUnicodeValueSlice(testArg)
	if err == nil {
		t.Errorf("[FAIL] Expected error for invalid hexadecimal string and received no error")
	}
}

func TestGetUnicodeValueSingleInvalidExcludeIncorrectLengthHexadecimal(t *testing.T) {
	testArg := "x20"
	_, err := getUnicodeValueSlice(testArg)
	if err == nil {
		t.Errorf("[FAIL] Expected error for invalid length hexadecimal string and received no error")
	}
}

func TestGetUnicodeValueRangeInvalidExcludeFirstValueNotHexadecimal(t *testing.T) {
	testArg := "xZZZZ-0020"
	_, err := getUnicodeValueSlice(testArg)
	if err == nil {
		t.Errorf("[FAIL] Expected error for invalid hexadecimal string and received no error")
	}
}

func TestGetUnicodeValueRangeInvalidExcludeLastValueNotHexadecimal(t *testing.T) {
	testArg := "x0020-ZZZZ"
	_, err := getUnicodeValueSlice(testArg)
	if err == nil {
		t.Errorf("[FAIL] Expected error for invalid hexadecimal string and received no error")
	}
}


tests/test_main.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import subprocess
import pytest


# Command line version, help, usage flag tests

def test_unilist_main_version_request_short():
    out = subprocess.check_output(
        "./unilist -v",
        stderr=subprocess.STDOUT,
        shell=True)

    assert "unilist v" in str(out)


def test_unilist_main_version_request_long():
    out = subprocess.check_output(
        "./unilist --version",
        stderr=subprocess.STDOUT,
        shell=True)

    assert "unilist v" in str(out)


def test_unilist_main_usage_request():
    out = subprocess.check_output(
        "./unilist --usage",
        stderr=subprocess.STDOUT,
        shell=True)

    assert "Usage:" in str(out)


def test_unilist_main_help_request_short():
    out = subprocess.check_output(
        "./unilist -h",
        stderr=subprocess.STDOUT,
        shell=True)

    assert "==========" in str(out)


def test_unilist_main_help_request_long():
    out = subprocess.check_output(
        "./unilist --help",
        stderr=subprocess.STDOUT,
        shell=True)

    assert "==========" in str(out)


# Command line validation tests


def test_unilist_main_too_few_args():
    with pytest.raises(subprocess.CalledProcessError):
        out = subprocess.check_output(
            "./unilist",
            stderr=subprocess.STDOUT,
            shell=True)


# Single Unicode code point argument tests

def test_unilist_main_single_valid_hexadecimal():
    out = subprocess.check_output(
        "./unilist 0020",
        stderr=subprocess.STDOUT,
        shell=True)

    assert out.decode('utf-8') == "U+0020"


def test_unilist_main_single_valid_hexadecimal_comma_delimited():
    out = subprocess.check_output(
        "./unilist -c 0020",
        stderr=subprocess.STDOUT,
        shell=True)

    assert out.decode('utf-8') == "U+0020"


def test_unilist_main_single_valid_hexadecimal_newline_delimited():
    out = subprocess.check_output(
        "./unilist -n 0020",
        stderr=subprocess.STDOUT,
        shell=True)

    assert out.decode('utf-8') == "U+0020"


def test_unilist_main_single_points_multiple_args_valid_hexadecimal():
    out = subprocess.check_output(
        "./unilist 0020 0021",
        stderr=subprocess.STDOUT,
        shell=True)

    assert out.decode('utf-8') == "U+0020 U+0021"


def test_unilist_main_single_points_multiple_args_valid_hexadecimal_comma_delimited():
    out = subprocess.check_output(
        "./unilist -c 0020 0021",
        stderr=subprocess.STDOUT,
        shell=True)

    assert out.decode('utf-8') == "U+0020,U+0021"


def test_unilist_main_single_points_multiple_args_valid_hexadecimal_newline_delimited():
    out = subprocess.check_output(
        "./unilist -n 0020 0021",
        stderr=subprocess.STDOUT,
        shell=True)

    assert out.decode('utf-8') == "U+0020\nU+0021"


def test_unilist_main_single_invalid_unicode_value():
    with pytest.raises(subprocess.CalledProcessError):
        out = subprocess.check_output(
            "./unilist ZZZZ",
            stderr=subprocess.STDOUT,
            shell=True)


def test_unilist_main_single_invalid_unicode_value_tooshort():
    with pytest.raises(subprocess.CalledProcessError):
        out = subprocess.check_output(
            "./unilist 20",
            stderr=subprocess.STDOUT,
            shell=True)


# Unicode code point range argument tests

def test_unilist_main_range_valid_hexadecimal():
    out = subprocess.check_output(
        "./unilist 0020-0022",
        stderr=subprocess.STDOUT,
        shell=True)

    assert out.decode('utf-8') == "U+0020 U+0021 U+0022"


def test_unilist_main_range_valid_hexadecimal_comma_delimited():
    out = subprocess.check_output(
        "./unilist -c 0020-0022",
        stderr=subprocess.STDOUT,
        shell=True)

    assert out.decode('utf-8') == "U+0020,U+0021,U+0022"


def test_unilist_main_range_valid_hexadecimal_newline_delimited():
    out = subprocess.check_output(
        "./unilist -n 0020-0022",
        stderr=subprocess.STDOUT,
        shell=True)

    assert out.decode('utf-8') == "U+0020\nU+0021\nU+0022"


def test_unilist_main_range_with_single_include_point():
    out = subprocess.check_output(
        "./unilist 0020-0022 0023",
        stderr=subprocess.STDOUT,
        shell=True)

    assert out.decode('utf-8') == "U+0020 U+0021 U+0022 U+0023"


def test_unilist_main_range_with_single_include_point_comma_delimited():
    out = subprocess.check_output(
        "./unilist -c 0020-0022 0023",
        stderr=subprocess.STDOUT,
        shell=True)

    assert out.decode('utf-8') == "U+0020,U+0021,U+0022,U+0023"


def test_unilist_main_range_with_single_include_point_newline_delimited():
    out = subprocess.check_output(
        "./unilist -n 0020-0022 0023",
        stderr=subprocess.STDOUT,
        shell=True)

    assert out.decode('utf-8') == "U+0020\nU+0021\nU+0022\nU+0023"


def test_unilist_main_range_invalid_start_unicode_value():
    with pytest.raises(subprocess.CalledProcessError):
        out = subprocess.check_output(
            "./unilist ZZZZ-0020",
            stderr=subprocess.STDOUT,
            shell=True)


def test_unilist_main_range_invalid_end_unicode_value():
    with pytest.raises(subprocess.CalledProcessError):
        out = subprocess.check_output(
            "./unilist 0020-ZZZZ",
            stderr=subprocess.STDOUT,
            shell=True)


def test_unilist_main_range_invalid_start_unicode_value_tooshort():
    with pytest.raises(subprocess.CalledProcessError):
        out = subprocess.check_output(
            "./unilist 20-0030",
            stderr=subprocess.STDOUT,
            shell=True)


def test_unilist_main_range_invalid_end_unicode_value_tooshort():
    with pytest.raises(subprocess.CalledProcessError):
        out = subprocess.check_output(
            "./unilist 0020-30",
            stderr=subprocess.STDOUT,
            shell=True)


# Unicode code point exclude tests

def test_unilist_main_exclude_single_value():
    out = subprocess.check_output(
        "./unilist 0020-0022 x0021",
        stderr=subprocess.STDOUT,
        shell=True)

    assert out.decode('utf-8') == "U+0020 U+0022"


def test_unilist_main_exclude_single_value_comma_delimited():
    out = subprocess.check_output(
        "./unilist -c 0020-0022 x0021",
        stderr=subprocess.STDOUT,
        shell=True)

    assert out.decode('utf-8') == "U+0020,U+0022"


def test_unilist_main_exclude_single_value_newline_delimited():
    out = subprocess.check_output(
        "./unilist -n 0020-0022 x0021",
        stderr=subprocess.STDOUT,
        shell=True)

    assert out.decode('utf-8') == "U+0020\nU+0022"


def test_unilist_main_exclude_range():
    out = subprocess.check_output(
        "./unilist 0020-0025 x0021-0024",
        stderr=subprocess.STDOUT,
        shell=True)

    assert out.decode('utf-8') == "U+0020 U+0025"


def test_unilist_main_exclude_range_comma_delimited():
    out = subprocess.check_output(
        "./unilist -c 0020-0025 x0021-0024",
        stderr=subprocess.STDOUT,
        shell=True)

    assert out.decode('utf-8') == "U+0020,U+0025"


def test_unilist_main_exclude_range_newline_delimited():
    out = subprocess.check_output(
        "./unilist -n 0020-0025 x0021-0024",
        stderr=subprocess.STDOUT,
        shell=True)

    assert out.decode('utf-8') == "U+0020\nU+0025"


def test_unilist_main_exclude_invalid_single_hexadecimal():
    with pytest.raises(subprocess.CalledProcessError):
        out = subprocess.check_output(
            "./unilist 0020-0025 xZZZZ",
            stderr=subprocess.STDOUT,
            shell=True)


def test_unilist_main_exclude_invalid_multiple_hexadecimal_start():
    with pytest.raises(subprocess.CalledProcessError):
        out = subprocess.check_output(
            "./unilist 0020-0025 xZZZZ-0030",
            stderr=subprocess.STDOUT,
            shell=True)


def test_unilist_main_exclude_invalid_multiple_hexadecimal_end():
    with pytest.raises(subprocess.CalledProcessError):
        out = subprocess.check_output(
            "./unilist 0020-0025 x0020-ZZZZ",
            stderr=subprocess.STDOUT,
            shell=True)


def test_unilist_main_exclude_invalid_single_hexadecimal_tooshort():
    with pytest.raises(subprocess.CalledProcessError):
        out = subprocess.check_output(
            "./unilist 0020-0025 x20",
            stderr=subprocess.STDOUT,
            shell=True)


def test_unilist_main_exclude_invalid_multiple_hexadecimal_start_tooshort():
    with pytest.raises(subprocess.CalledProcessError):
        out = subprocess.check_output(
            "./unilist 0020-0025 x20-0030",
            stderr=subprocess.STDOUT,
            shell=True)


def test_unilist_main_exclude_invalid_multiple_hexadecimal_end_tooshort():
    with pytest.raises(subprocess.CalledProcessError):
        out = subprocess.check_output(
            "./unilist 0020-0025 x0020-30",
            stderr=subprocess.STDOUT,
            shell=True)




stringutils_test.go
package main

import (
	"testing"
)

func TestContainsStringValid(t *testing.T) {
	testSlice := []string{"test", "another", "last"}
	if !containsString(testSlice, "test") {
		t.Errorf("[FAIL] Expected test for string that is present to return true and it returned false")
	}
	if !containsString(testSlice, "another") {
		t.Errorf("[FAIL] Expected test for string that is present to return true and it returned false")
	}
	if !containsString(testSlice, "last") {
		t.Errorf("[FAIL] Expected test for string that is present to return true and it returned false")
	}
}

func TestContainsStringInvalid(t *testing.T) {
	testSlice := []string{"test", "another", "last"}
	if containsString(testSlice, "bogus") {
		t.Errorf("[FAIL] Expected test for string that is not present to return false and it returned true")
	}
}

func TestGetCommaDelimitedString(t *testing.T) {
	testSlice := []string{"test", "another", "last"}
	responseString := getCommaDelimitedString(testSlice)
	if responseString != "test,another,last" {
		t.Errorf("[FAIL] Expected test to return comma delimited string and instead returned '%s'", responseString)
	}
}

func TestGetNewlineDelimitedString(t *testing.T) {
	testSlice := []string{"test", "another", "last"}
	responseString := getNewlineDelimitedString(testSlice)
	if responseString != "test\nanother\nlast" {
		t.Errorf("[FAIL] Expected test to return newline delimited string and instead returned '%s'", responseString)
	}
}

func TestGetSpaceDelimitedString(t *testing.T) {
	testSlice := []string{"test", "another", "last"}
	responseString := getSpaceDelimitedString(testSlice)
	if responseString != "test another last" {
		t.Errorf("[FAIL] Expected test to return space delimited string and instead returned '%s'", responseString)
	}
}


parser.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func getUnicodeValueSlice(arg string) ([]string, error) {
	var returnSlice []string
	// strip the case insensitive 'x' value from exclude request arguments (detected in calling code)
	if strings.HasPrefix(arg, "x") || strings.HasPrefix(arg, "X") {
		arg = arg[1:]
	}
	if strings.Contains(arg, "-") {
		tempUniList := strings.Split(arg, "-")
		if len(tempUniList[0]) < 4 {
			// handle argument that is not formatted as unicode hexadecimal value
			lenErr := fmt.Errorf("the argment '%s' is not properly formatted as a Unicode hexadecimal value", tempUniList[0])
			return returnSlice, lenErr
		}
		if len(tempUniList[1]) < 4 {
			// handle argument that is not formatted as unicode hexadecimal value
			lenErr := fmt.Errorf("the argment '%s' is not properly formatted as a Unicode hexadecimal value", tempUniList[1])
			return returnSlice, lenErr
		}
		startInt, startParseErr := strconv.ParseInt(tempUniList[0], 16, 32)
		endInt, endParseErr := strconv.ParseInt(tempUniList[1], 16, 32)
		if startParseErr != nil {
			return returnSlice, startParseErr
		}
		if endParseErr != nil {
			return returnSlice, endParseErr
		}
		for i := startInt; i <= endInt; i++ {
			returnSlice = append(returnSlice, fmt.Sprintf("%U", i))
		}
	} else {
		if len(arg) > 3 {
			anInt, parseErr := strconv.ParseInt(arg, 16, 32)
			if parseErr != nil {
				return returnSlice, parseErr
			}
			returnSlice = append(returnSlice, fmt.Sprintf("%U", anInt))

		} else {
			// handle argument that is not formatted as unicode hexadecimal value
			lenErr := fmt.Errorf("the argment '%s' is not properly formatted as a Unicode hexadecimal value", arg)
			return returnSlice, lenErr
		}
	}
	return returnSlice, nil
}


unilist.go
// unilist is a command line executable that generates lists of Unicode code points
package main

import (
	"flag"
	"fmt"
	"os"
	"strings"
)

const (
	version = "0.3.0"

	usage = `Usage: unilist (options) [arg 1]...[arg n]
`

	help = `=================================================
 unilist
 Copyright 2018 Christopher Simpkins
 MIT License

 Source: https://github.com/source-foundry/unilist
=================================================
 Usage:
   $ unilist (options) [arg 1]...[arg n]

 Default:
   Write a list of Unicode code point values delimited by space glyphs
   to the standard output stream

 Options:
  -c, --comma          Use comma delimiter
  -h, --help           Application help
  -n, --newline        Use newline delimiter
      --usage          Application usage
  -v, --version        Application version

 Notes:
   Define arg values as single Unicode hexadecimal code point
   values or as ranges that are indicated with a '-' between 
   values (e.g., 0020 or 0020-0024). 

   Exclude hexadecimal values from output by using a 
   case-insensitive 'x' as the first character of the single
   Unicode code point arg or range arg (e.g., x0020 or x0020-0024)
`
)

var versionShort, versionLong, helpShort, helpLong, usageLong *bool
var useCommasLong, useCommasShort, useNewLinesLong, useNewLinesShort *bool

func init() {
	// define available command line flags
	versionShort = flag.Bool("v", false, "Application version")
	versionLong = flag.Bool("version", false, "Application version")
	helpShort = flag.Bool("h", false, "Help")
	helpLong = flag.Bool("help", false, "Help")
	usageLong = flag.Bool("usage", false, "Usage")
	useCommasLong = flag.Bool("comma", false, "Comma delimiters in lists")
	useCommasShort = flag.Bool("c", false, "Comma delimiters in lists")
	useNewLinesLong = flag.Bool("newline", false, "Newline delimiters in lists")
	useNewLinesShort = flag.Bool("n", false, "Newline delimiters in lists")
}

func main() {
	flag.Parse()

	// parse command line flags and handle them
	switch {
	case *versionShort, *versionLong:
		fmt.Printf("unilist v%s\n", version)
		os.Exit(0)
	case *helpShort, *helpLong:
		fmt.Println(help)
		os.Exit(0)
	case *usageLong:
		fmt.Println(usage)
		os.Exit(0)
	}

	if len(flag.Args()) == 0 {
		os.Stderr.WriteString("[Error] Please include at least one Unicode code point range argument in your command.\n")
		os.Exit(1)
	}

	var keepUniList []string
	var removeUniList []string

	// create exclude Unicode value slice
	for _, remArg := range flag.Args() {
		if strings.HasPrefix(remArg, "x") || strings.HasPrefix(remArg, "X") {
			tempRemoveSlice, err := getUnicodeValueSlice(remArg)
			if err != nil {
				os.Stderr.WriteString(fmt.Sprintf("[Error] %v", err))
				os.Exit(1)
			}

			removeUniList = append(removeUniList, tempRemoveSlice...)
		}
	}

	// create include Unicode value slice, taking into account above excluded values
	for _, keepArg := range flag.Args() {
		if !strings.HasPrefix(keepArg, "x") || strings.HasPrefix(keepArg, "X") {
			tempKeepSlice, err := getUnicodeValueSlice(keepArg)
			if err != nil {
				os.Stderr.WriteString(fmt.Sprintf("[Error] %v", err))
				os.Exit(1)
			}

			for _, keepValue := range tempKeepSlice {
				if !containsString(removeUniList, keepValue) {
					// only insert into the keep values slice if it is not in the exclude values slice
					keepUniList = append(keepUniList, keepValue)
				}
			}
		}
	}

	// format the standard out string with proper default or command line defined delimiter
	if *useCommasShort || *useCommasLong {
		fmt.Print(getCommaDelimitedString(keepUniList))
	} else if *useNewLinesShort || *useNewLinesLong {
		fmt.Print(getNewlineDelimitedString(keepUniList))
	} else {
		fmt.Print(getSpaceDelimitedString(keepUniList))
	}
}
